jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java

```diff
@@ -115,13 +115,53 @@ public ServerPool(final Set<String> servers) {
         }
 
         public String getNextServer() {
-            if (serverIterator.hasNext()) return serverIterator.next();
+            String server = serversRing.next();
+            if(server != null) return server;
             else throw new NoServerConfiguredException("No Server is assigned to client to connect");
         }
 
         public int getSize() {
             return servers.size();
         }
+    }
+
+    private static final class ImmutableCircularList<E> {
+
+        final int size;
+        Node<E> current;
 
+        private static final class Node<E> {
+            E item;
+            Node<E> next;
+
+            Node(E data, Node<E> next) {
+                this.item = data;
+                this.next = next;
+            }
+        }
+
+        public ImmutableCircularList(Collection<? extends E> c) {
+            size = c.size();
+            if (size == 0)
+                return;
+
+            Iterator<?> iter = c.iterator();
+            Node<E> head = new Node<E>((E) iter.next(), null);
+            current = head;
+            while (iter.hasNext()) {
+                Node<E> nextNode = new Node<E>((E) iter.next(), null);
+                current.next = nextNode;
+                current = nextNode;
+            }
+            current.next = head;
+            current = head;
+        }
+
+        public E next() {
+            if(current == null) return null;
+            E data = current.item;
```

Comment:
This is a slick solution. Doesn't it add a performance hit though? As for the skipping or getting the same server for consecutive calls, I don't know the internals of Elastic but would it cause a performance hit? By synchronising the `getNextServer` method then these problems would vanish too. Moreover, depending on the use case sync could outperform atomic or vice versa according to this http://stackoverflow.com/questions/3556283/in-java-what-is-the-performance-of-atomicinteger-compareandset-versus-synchron. What are your thoughts?